File: scripts/day6-part2-simple.js
const fs = require("fs");

function solvePart2(input) {
  const lines = input.split("\n");
  const height = lines.length;
  const width = Math.max(...lines.map((l) => l.length));

  const normalizedLines = lines.map((line) => line.padEnd(width, " "));

  const problems = [];
  let col = width - 1;

  while (col >= 0) {
    let hasNonSpace = false;
    for (let row = 0; row < height; row++) {
      if (normalizedLines[row][col] !== " ") {
        hasNonSpace = true;
        break;
      }
    }

    if (!hasNonSpace) {
      col--;
      continue;
    }

    let startCol = col;
    while (startCol > 0) {
      let isAllSpace = true;
      for (let row = 0; row < height; row++) {
        if (normalizedLines[row][startCol - 1] !== " ") {
          isAllSpace = false;
          break;
        }
      }
      if (isAllSpace) break;
      startCol--;
    }

    const numbers = [];
    let operator = null;

    for (let c = col; c >= startCol; c--) {
      let digitStr = "";

      for (let row = 0; row < height; row++) {
        const char = normalizedLines[row][c];
        if (char === "+" || char === "*") {
          operator = char;
        } else if (char !== " ") {
          digitStr += char;
        }
      }

      if (digitStr.length > 0) {
        numbers.push(parseInt(digitStr));
      }
    }

    if (operator && numbers.length > 0) {
      problems.push({ numbers, operator });
    }
    col = startCol - 1;
  }

  let grandTotal = 0;

  for (const problem of problems) {
    let result;
    if (problem.operator === "+") {
      result = problem.numbers.reduce((a, b) => a + b, 0);
    } else {
      result = problem.numbers.reduce((a, b) => a * b, 1);
    }
    grandTotal += result;
  }

  return grandTotal;
}

const input = fs.readFileSync("../inputs/day-6-input.txt", "utf-8").trim();

const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart2(input);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Day 6 Part 2 - Simple)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMicros} microseconds (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);

File: scripts/day6-part2-simple.py
import time

def solve_part2(input_data):
    lines = input_data.split('\n')
    height = len(lines)
    width = max(len(line) for line in lines)
    
    normalized_lines = [line.ljust(width) for line in lines]
    
    problems = []
    col = width - 1
    
    while col >= 0:
        has_non_space = False
        for row in range(height):
            if normalized_lines[row][col] != ' ':
                has_non_space = True
                break
        
        if not has_non_space:
            col -= 1
            continue
        
        start_col = col
        while start_col > 0:
            is_all_space = True
            for row in range(height):
                if normalized_lines[row][start_col - 1] != ' ':
                    is_all_space = False
                    break
            if is_all_space:
                break
            start_col -= 1
        
        numbers = []
        operator = None
        
        for c in range(col, start_col - 1, -1):
            digit_str = ""
            
            for row in range(height):
                char = normalized_lines[row][c]
                if char in ('+', '*'):
                    operator = char
                elif char != ' ':
                    digit_str += char
            
            if digit_str:
                numbers.append(int(digit_str))
        
        if operator and numbers:
            problems.append({'numbers': numbers, 'operator': operator})
        
        col = start_col - 1
    
    grand_total = 0
    
    for problem in problems:
        if problem['operator'] == '+':
            result = sum(problem['numbers'])
        else:
            result = 1
            for num in problem['numbers']:
                result *= num
        grand_total += result
    
    return grand_total

# Read input
with open("../inputs/day-6-input.txt", "r") as f:
    input_data = f.read().strip()

# Benchmark
iterations = 1
start_time = time.perf_counter()

for _ in range(iterations):
    result = solve_part2(input_data)

end_time = time.perf_counter()
total_time = end_time - start_time
avg_time_us = (total_time / iterations) * 1_000_000

print(f"Python (Day 6 Part 2 - Simple)")
print(f"Answer: {result}")
print(f"Average time: {avg_time_us:.3f} microseconds ({iterations} iterations)")
print(f"Total time: {total_time * 1000:.2f} ms")
